dali/test/python/test_RN50_external_source_parallel.py

```diff
@@ -0,0 +1,363 @@
+
+    return args
+
+
+def iteration_test(args):
+
+    for pipe_factory in TEST_DATA:
+        pipe = pipe_factory(
+            batch_size=args.batch_size,
+            num_threads=args.workers,
+            device_id=args.assign_gpu,
+            data_path=args.data_path,
+            prefetch=args.prefetch,
+            reader_queue_depth=args.reader_queue_depth,
+            py_workers_init=args.worker_init,
+            py_workers_num=args.py_workers
+            )
+        pipe.build()
+
+        samples_no = pipe.epoch_size("Reader")
+        if args.benchmark_iters is None:
+            expected_iters = samples_no // args.batch_size + (samples_no % args.batch_size != 0)
+        else:
+            expected_iters = args.benchmark_iters
+
+        print("RUN {}".format(pipe_factory.__name__))
+        for i in range(args.epochs):
+            if i == 0:
+                print("Warm up")
+            else:
+                print("Test run " + str(i))
+            data_time = AverageMeter()
+            end = time.time()
+            for j in range(expected_iters):
+                try:
+                    pipe.run()
+                except StopIteration:
+                    assert j == expected_iters - 1
+                    break
+                data_time.update(time.time() - end)
+                if j % 10 == 0:
+                    print("{} {}/ {}, avg time: {} [s], worst time: {} [s], speed: {} [img/s]".format(
+                            pipe_factory.__name__,
+                            j,
+                            expected_iters,
+                            data_time.avg,
+                            data_time.max_val,
+                            args.batch_size / data_time.avg,
+                    ))
+                end = time.time()
+            pipe.reset()
+
+        print("OK {}".format(pipe_factory.__name__))
+
+
+def training_test(args):
+
+    import torch
+    import torch.nn as nn
+    import torch.optim
+    import torchvision.models as models
+    from nvidia.dali.plugin.pytorch import DALIClassificationIterator
+
+    for pipe_factory in TEST_DATA:
+        pipe = pipe_factory(
+            batch_size=args.batch_size,
+            num_threads=args.workers,
+            device_id=args.assign_gpu,
+            data_path=args.data_path,
+            prefetch=args.prefetch,
+            reader_queue_depth=args.reader_queue_depth,
+            py_workers_init=args.worker_init,
+            py_workers_num=args.py_workers
+            )
+        pipe.build()
+
+        model =  models.resnet18().cuda()
+        model.train()
+        loss_fun = nn.CrossEntropyLoss().cuda()
+        lr = 0.1 * args.batch_size / 256
+        optimizer = torch.optim.SGD(
+            model.parameters(),
+            lr,
+            momentum=0.9,
+        )
+
+        samples_no = pipe.epoch_size("Reader")
```

Comment:
What we get for ExternalSource here?

Reply:
Good question.

Reply:
ExternalSourcePipeline has overloaded epoch_size that returns the size of the dataset as well.